tests/test-chat-template.cpp

```diff
@@ -49,6 +49,8 @@ int main(void) {
         "{{ bos_token }}{% if messages[0]['role'] == 'system' %}{% set loop_messages = messages[1:] %}{% set system_message = messages[0]['content'] %}{% elif false == true %}{% set loop_messages = messages %}{% set system_message = 'You are Command-R, a brilliant, sophisticated, AI-assistant trained to assist human users by providing thorough responses. You are trained by Cohere.' %}{% else %}{% set loop_messages = messages %}{% set system_message = false %}{% endif %}{% if system_message != false %}{{ '<|START_OF_TURN_TOKEN|><|SYSTEM_TOKEN|>' + system_message + '<|END_OF_TURN_TOKEN|>' }}{% endif %}{% for message in loop_messages %}{% if (message['role'] == 'user') != (loop.index0 % 2 == 0) %}{{ raise_exception('Conversation roles must alternate user/assistant/user/assistant/...') }}{% endif %}{% set content = message['content'] %}{% if message['role'] == 'user' %}{{ '<|START_OF_TURN_TOKEN|><|USER_TOKEN|>' + content.strip() + '<|END_OF_TURN_TOKEN|>' }}{% elif message['role'] == 'assistant' %}{{ '<|START_OF_TURN_TOKEN|><|CHATBOT_TOKEN|>'  + content.strip() + '<|END_OF_TURN_TOKEN|>' }}{% endif %}{% endfor %}{% if add_generation_prompt %}{{ '<|START_OF_TURN_TOKEN|><|CHATBOT_TOKEN|>' }}{% endif %}",
         // Llama-3
         "{% set loop_messages = messages %}{% for message in loop_messages %}{% set content = '<|start_header_id|>' + message['role'] + '<|end_header_id|>\n\n'+ message['content'] | trim + '<|eot_id|>' %}{% if loop.index0 == 0 %}{% set content = bos_token + content %}{% endif %}{{ content }}{% endfor %}{{ '<|start_header_id|>assistant<|end_header_id|>\n\n' }}",
+        // Phi-3
+        "{{ bos_token }}{% for message in messages %}{{'<|' + message['role'] + '|>' + ' ' + message['content'] + '<|end|> ' }}{% endfor %}{% if add_generation_prompt %}{{ '<|assistant|> ' }}{% else %}{{ eos_token }}{% endif %}"
```

Comment:
@tristandruyen This template uses a space character `' '` before each special token, but on your cpp version, you used new line `"\n"`. Can you check if it is correct?

Reply:
According to the huggingface README's there should be a `\n` after `<|user|>` https://huggingface.co/microsoft/Phi-3-mini-128k-instruct/blob/main/README.md#chat-format

Reply:
While they do not explicitly state that there should be a `\n` after `<|assistant|>` 99% of the time the model generates a `\n` as a first token when it is left out...

Reply:
> According to the Chat Format documented in the huggingface README's there should be a \n after <|user|> and <|end|>

It's best to look at the code directly rather than README. Sometimes human copy-paste can have mistakes.

On the same repo that you linked to, you can see the jinja template code here:

https://huggingface.co/microsoft/Phi-3-mini-128k-instruct/blob/main/tokenizer_config.json#L119

This file have `'<|user|>' + '\n'` which confirms that there should be a new line after special tokens.

However, that does not explain why your jinja version does not have it. Remember, jinja templates are deterministic, meaning your cpp code need to follow 100% what is being coded inside the template. This mean we can fix this PR in one of 2 ways: (1) correct the jinja template to have `\n` or (2) correct the cpp implementation output `' '` instead of new line.

Reply:
Seems there are even more version of the template, I guess i had an outdated tokeinzer_config.json when making my GGUF's, because I just took the jinja template from the gguf metadata of my recent phi3 medium quants...

Reply:
Looking at the commit history, they never seem to use space instead of new line. I think maybe the code you used to dump the gguf metadata replaced new line with a space?

IMO I don't believe that there is more than one version of the template. Think of it: why would microsoft re-train or re-fine-tune the whole model just to introduce a small change in chat template?

Reply:
Edit: there are indeed 2 versions, but they all use new line:
- One with `'<|' + message['role'] + '|>' + '\n'`: https://huggingface.co/microsoft/Phi-3-small-128k-instruct/blob/main/tokenizer_config.json#L10
- And another one with explicitly `<|user|>`: https://huggingface.co/microsoft/Phi-3-mini-128k-instruct/blob/main/tokenizer_config.json#L119

I think it's best to somehow scan all template in existing [phi-3 family](https://huggingface.co/collections/microsoft/phi-3-6626e15e9585a200d2d761e3), then make a fix PR. Without careful research, I doubt that there will be more confusions in the future

Reply:
> I think I found the reason, the huggingface GUI for displaying GGUF metadata, seems to replace \n with whitespace for some weird reason....

I'd agree that there is a bad design in the jinja templates (in general). Since `'\n'` is wrapped inside a json string, it should be converted to `" '\\n' "`. But for some reasons, jinja doesn't care about this mistake (that's why it get converted to space in the gguf viewer on huggingface).

Unfortunately, this is not something we could fix on our side.

Reply:
I updated the PR :)